example_allocators/compartment_alloc/compartment_alloc.c

```diff
@@ -0,0 +1,91 @@
+#include <cheriintrin.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include "compartment_alloc.h"
+#include "utils.h"
+
+heap_compartment compartments[maxCompartments];
+int numCompartments = 0;
+
+/*
+ * Initialise a compartment with `size_in_bytes` space, where all future
+ * allocations will be constrained by `dc` as if it was the DDC when allocation
+ * occurred.
+ */
+void
+*__capability init_compartment(int size_in_bytes, void *__capability dc)
+{
+    // We impose a limit on our # of compartments
+    if (numCompartments + 1 == maxCompartments) {
+        return NULL;
+    }
+
+    // Allocate memory for this compartment
+    void *buf = mmap(NULL, size_in_bytes, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+    if (buf == MAP_FAILED)
+    {
+        perror("error in initial mem allocation");
+        exit(-1);
+    }
+
+    // Create the new compartment
+    compartments[numCompartments] = (heap_compartment){
+        buf,                     // Space on the heap
+        0,                       // bytes allocated (currently none!)
+        size_in_bytes,           // Our maximum size, to check when allocating
+        dc,                      // The data capability to allocate against (like a DDC)
+        sealed_reference(buf)    // An identifier which allows a holder to allocate in this component
+    };
+
+    // Return the sealed capability which acts as a token of authentication for a component.
+    return compartments[numCompartments++].identifier;
```

Comment:
Is this meant to also increment `numCompartments`? Personally I would put it on a separate line, as this effect might be overlooked, but I don't feel too strongly about it.

Reply:
Yes! And I understand — made the change, fixed in [b84412b](https://github.com/capablevms/cheri-examples/pull/88/commits/b84412b1a51c116b237d5ebe1b00e636bbd2cf05).